hal-sim/hal_impl/functions.py

```diff
@@ -856,7 +856,24 @@ def freeDIOPort(dioPortHandle):
 
 def allocateDigitalPWM(status):
     status.value = 0
-    return types.DigitalPWMHandle()
+    handle = 0
+    dio = None
+    for i, port in enumerate(hal_data["dio"]):
+        if i >= kNumDigitalHeaders:
+            if not hal_data["mxp"][i]["initialized"]:
+                dio = _initport('dio', i, status)
+                hal_data["mxp"][i]["initialized"] = True
+                handle = i
+                break
+        if not port['initialized']:
+            dio = _initport('dio', i, status)
+            handle = i
+            break
+
+    if dio is None:
+        return 0
+
+    return types.DigitalPWMHandle(getPort(handle))
```

Comment:
@virtuald / @auscompgeek can you confirm that this is correct logically and the correct way to go about this? I think I interpreted the C code correctly, but not 100% sure.

Reply:
I'm having trouble figuring out how you got to this implementation. `HAL_AllocateDigitalPWM` is _very_ different…